### PR TITLE
RFD 0169 - Application MFA Sessions

### PR DESCRIPTION
**Pull request by Joerger**
Updates https://github.com/gravitational/teleport/issues/20634

### CODE REVIEW

**Review by rosstimothy — COMMENTED** [review]
What kind of implications does this have for vnet?

File: rfd/0168-app-mfa-sessions.md [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
++ MFAAuthenticateResponse MFAResponse = 8 [(gogoproto.jsontag) = "mfa_response,omitempty"];
+}
+```
+
+The TTL of app sessions created with `CreateAppSession` will default to the TTL
+of the user's current certificates. This is the existing behaviour.
+
+Note: this flow will be used by WebUI clients connecting to an Application.
+This also includes Teleport Connect which uses deep links to open Applications
+in the WebUI.
```
[the remark this answers is not in the document]
Comment (Joerger) [reply rfd/0168-app-mfa-sessions.md]: This section is referring to Teleport Connect using a WebUI deep link to connect to an App, and the second _should_ be referring to Teleport Connect opening a local app proxy. I'll rework these sections to make it make sense.

**Comment by Joerger** [comment]
> What kind of implications does this have for vnet?

We talked about this in our 1x1 briefly but I looked into the RFD VNet to learn more. IIUC while it is a big change to App Access it does not change the underlying mechanism used to create app sessions and app certificates, only the DNS routing used to connect to the App Service. This would all take place after the MFA checks and shouldn't interfere in general.

Depending on the Teleport Connect implementation, there may be issue with renewing MFA verified certs for Apps already registered in VNet, but if the App Gateway is implemented to handle certificate reissuing and MFA checks it should be fine.

@ravicious do you suspect there are any implications I'm missing? This may be something that we'll need to test after both implementations are complete and fix it if it doesn't work out of the box.

**Comment by ravicious** [comment]
I skimmed through this RFD and I think your comment is right, VNet should work with per-session MFA out of the box. The VNet manager accepts a `TeleportClient` and [uses it to call `CreateAppSession` and `ReissueUserCerts`](https://github.com/gravitational/teleport/blob/39ad75b5e51ef7480072eacd4b6e791a871f8843/lib/vnet/app.go#L151-L210). So as long as the per-session MFA for apps is going to use the MFA prompt, then Connect should have no issues with handling it, similar to MFA for kube proxies or admin actions.

Overall VNet will behave very similar to `tsh proxy app` in that regard, since as you said it's pretty much just that with a different routing to the app.

**Review by jentfoo — APPROVED** [review]

**Review by rosstimothy — APPROVED** [review]

**Comment by Joerger** [comment]
@xinding33 @klizhentas If you have the chance could you give this RFD a quick look? From a UX perspective there isn't much different from the DB and Kube implementations.

**Comment by Joerger** [comment]
@nklaassen @zmb3 Just need one more review, thanks!

**Review by nklaassen — APPROVED** [review]

**Pull request merged**